### ISSUE TITLE
apart: hangs or takes too long

### ISSUE THREAD

**Issue by proy87** [sympy/sympy#18531]
Usin live.sympy.org it appears that `((x**2 + 1)**3/((x - 1)**2*(x + 1)**2*(-x**2 + 2*x + 1)*(x**2 + 2*x - 1))).apart(x=x,extension={sqrt(2)})` hangs.

**Comment by gschintgen** [comment]
With (almost) master on a rather weak machine:
```
In [1]: time ((x**2 + 1)**3/((x - 1)**2*(x + 1)**2*(-x**2 + 2*x + 1)*(x**2 + 2*x - 1))).apart(x=x,extension={sqrt(2)})
CPU times: user 6min 7s, sys: 67.4 ms, total: 6min 7s
Wall time: 6min 7s
Out[1]:
      1                1                1                1              1            1
────────────── - ────────────── + ────────────── - ────────────── + ────────── + ──────────
2⋅(x - √2 + 1)   2⋅(x - √2 - 1)   2⋅(x + 1 + √2)   2⋅(x - 1 + √2)            2            2
                                                                    2⋅(x + 1)    2⋅(x - 1)
```
So, yes, it's slow.

AFAIK there's also a very short timeout on Sympy Live.

**Comment by oscarbenjamin** [comment]
I profiled this with
```console
$ python -m cProfile -s cumulative t.py > t.py.time
```
The result:
```
         648705183 function calls (598362868 primitive calls) in 499.909 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    800/1    0.026    0.000  500.923  500.923 {built-in method builtins.exec}
        1    0.000    0.000  500.923  500.923 t.py:1(<module>)
        1    0.000    0.000  500.098  500.098 expr.py:3551(apart)
        1    0.000    0.000  500.098  500.098 decorator.py:18(threaded_func)
        1    0.000    0.000  500.098  500.098 partfrac.py:14(apart)
        1    0.001    0.001  500.022  500.022 partfrac.py:151(apart_undetermined_coeffs)
        1    0.002    0.002  499.398  499.398 solvers.py:472(solve)
        1    0.007    0.007  499.256  499.256 solvers.py:1788(_solve_system)
        1    0.000    0.000  498.975  498.975 solvers.py:2257(solve_linear_system)
3926124/2421337    2.771    0.000  311.598    0.000 assumptions.py:258(getit)
321792/7886    1.284    0.000  309.793    0.039 assumptions.py:270(_ask)
1642/1101    0.028    0.000  299.528    0.272 expr.py:879(_eval_is_extended_positive_negative)
        8    0.001    0.000  293.063   36.633 simplify.py:395(simplify)
    12643    0.080    0.000  207.627    0.016 evalf.py:1373(evalf)
10424853/12643   30.932    0.000  207.379    0.016 evalf.py:1310(evalf)
1127325/7201    9.512    0.000  207.247    0.029 evalf.py:511(evalf_add)
1127389/7265    1.965    0.000  206.585    0.028 evalf.py:526(<listcomp>)
1163498/40178   21.761    0.000  206.193    0.005 evalf.py:553(evalf_mul)
        1    0.001    0.001  203.299  203.299 solvers.py:2630(inv_quick)
457114/399058    7.024    0.000  189.463    0.000 operations.py:30(__new__)
```
So the time is spent in `solve_linear_system`.

The call to `solve_linear_system` is:
```python
from sympy import *
M = Matrix([
    [1, 1, 1, 1, 1, 0, 1, 0, 0],
    [1 + sqrt(2), -1 + sqrt(2), 1 - sqrt(2), -sqrt(2) - 1, 1, 1, -1, 1, 1],
    [-5 + 2*sqrt(2), -5 - 2*sqrt(2), -5 - 2*sqrt(2), -5 + 2*sqrt(2), -7, 2, -7, -2, 0],
    [-3*sqrt(2) - 1, 1 - 3*sqrt(2), -1 + 3*sqrt(2), 1 + 3*sqrt(2), -7, -5, 7, -5, 3],
    [7 - 4*sqrt(2), 4*sqrt(2) + 7, 4*sqrt(2) + 7, 7 - 4*sqrt(2), 7, -12, 7, 12, 0],
    [-1 + 3*sqrt(2), 1 + 3*sqrt(2), -3*sqrt(2) - 1, 1 - 3*sqrt(2), 7, -5, -7, -5, 3],
    [-3 + 2*sqrt(2), -3 - 2*sqrt(2), -3 - 2*sqrt(2), -3 + 2*sqrt(2), -1, 2, -1, -2, 0],
    [1 - sqrt(2), -sqrt(2) - 1, 1 + sqrt(2), -1 + sqrt(2), -1, 1, 1, 1, 1]
    ])
print(solve_linear_system(M, *symbols('x:8')))
```
That matrix is:
```julia
In [2]: M                                                                                                                                      
Out[2]: 
⎡    1          1          1          1      1    0   1   0   0⎤
⎢                                                              ⎥
⎢ 1 + √2     -1 + √2    1 - √2     -√2 - 1   1    1   -1  1   1⎥
⎢                                                              ⎥
⎢-5 + 2⋅√2  -5 - 2⋅√2  -5 - 2⋅√2  -5 + 2⋅√2  -7   2   -7  -2  0⎥
⎢                                                              ⎥
⎢-3⋅√2 - 1  1 - 3⋅√2   -1 + 3⋅√2  1 + 3⋅√2   -7  -5   7   -5  3⎥
⎢                                                              ⎥
⎢7 - 4⋅√2   4⋅√2 + 7   4⋅√2 + 7   7 - 4⋅√2   7   -12  7   12  0⎥
⎢                                                              ⎥
⎢-1 + 3⋅√2  1 + 3⋅√2   -3⋅√2 - 1  1 - 3⋅√2   7   -5   -7  -5  3⎥
⎢                                                              ⎥
⎢-3 + 2⋅√2  -3 - 2⋅√2  -3 - 2⋅√2  -3 + 2⋅√2  -1   2   -1  -2  0⎥
⎢                                                              ⎥
⎣ 1 - √2     -√2 - 1    1 + √2     -1 + √2   -1   1   1   1   1⎦
```
This kind of thing should not be slow (#18445)

**Comment by oscarbenjamin** [comment]
This can be fixed by #4949 sinve rref no longer has trouble with this matrix after #18572 
```python
In [1]: M = Matrix([ 
   ...:     [1, 1, 1, 1, 1, 0, 1, 0, 0], 
   ...:     [1 + sqrt(2), -1 + sqrt(2), 1 - sqrt(2), -sqrt(2) - 1, 1, 1, -1, 1, 1], 
   ...:     [-5 + 2*sqrt(2), -5 - 2*sqrt(2), -5 - 2*sqrt(2), -5 + 2*sqrt(2), -7, 2, -7, -2, 0], 
   ...:     [-3*sqrt(2) - 1, 1 - 3*sqrt(2), -1 + 3*sqrt(2), 1 + 3*sqrt(2), -7, -5, 7, -5, 3], 
   ...:     [7 - 4*sqrt(2), 4*sqrt(2) + 7, 4*sqrt(2) + 7, 7 - 4*sqrt(2), 7, -12, 7, 12, 0], 
   ...:     [-1 + 3*sqrt(2), 1 + 3*sqrt(2), -3*sqrt(2) - 1, 1 - 3*sqrt(2), 7, -5, -7, -5, 3], 
   ...:     [-3 + 2*sqrt(2), -3 - 2*sqrt(2), -3 - 2*sqrt(2), -3 + 2*sqrt(2), -1, 2, -1, -2, 0], 
   ...:     [1 - sqrt(2), -sqrt(2) - 1, 1 + sqrt(2), -1 + sqrt(2), -1, 1, 1, 1, 1] 
   ...:     ])                                                                                                                                                

In [2]: %time M.rref()                                                                                                                                        
CPU times: user 491 ms, sys: 3.64 ms, total: 495 ms
Wall time: 495 ms
Out[2]: 
⎛⎡1  0  0  0  0  0  0  0  1/2 ⎤                          ⎞
⎜⎢                            ⎥                          ⎟
⎜⎢0  1  0  0  0  0  0  0  -1/2⎥                          ⎟
⎜⎢                            ⎥                          ⎟
⎜⎢0  0  1  0  0  0  0  0  1/2 ⎥                          ⎟
⎜⎢                            ⎥                          ⎟
⎜⎢0  0  0  1  0  0  0  0  -1/2⎥                          ⎟
⎜⎢                            ⎥, (0, 1, 2, 3, 4, 5, 6, 7)⎟
⎜⎢0  0  0  0  1  0  0  0   0  ⎥                          ⎟
⎜⎢                            ⎥                          ⎟
⎜⎢0  0  0  0  0  1  0  0  -1/2⎥                          ⎟
⎜⎢                            ⎥                          ⎟
⎜⎢0  0  0  0  0  0  1  0   0  ⎥                          ⎟
⎜⎢                            ⎥                          ⎟
⎝⎣0  0  0  0  0  0  0  1  -1/2⎦                          ⎠
```
